libff/algebra/curves/tests/test_groups.cpp

```diff
@@ -317,6 +317,53 @@ void test_bls12_377()
     ASSERT_EQ(bls12_377_G2::zero(), z);
 }
 
+void test_bls12_381()
+{
+    using GroupT = typename std::decay<G1<bls12_381_pp>>::type;
+    // test that some elements e.g. 1,-1,2,random satisfy the curve
+    // equation Y^2 = X^3 + a X + b P = 1
+    // P = 1
+    {
+        GroupT P = GroupT::one();
+        P.to_affine_coordinates();
+        using Fq = typename std::decay<decltype(P.X)>::type;
+        Fq LHS = (P.Y * P.Y);
+        Fq RHS =
+            ((P.X * P.X * P.X) + (GroupT::coeff_a * P.X) + GroupT::coeff_b);
+        ASSERT_EQ(LHS, RHS);
+    }
+    // P = -1
+    {
+        GroupT P = bigint<1>(-1l) * GroupT::one();
+        P.to_affine_coordinates();
+        using Fq = typename std::decay<decltype(P.X)>::type;
+        Fq LHS = (P.Y * P.Y);
+        Fq RHS =
+            ((P.X * P.X * P.X) + (GroupT::coeff_a * P.X) + GroupT::coeff_b);
+        ASSERT_EQ(LHS, RHS);
+    }
+    // P = 2
+    {
+        GroupT P = bigint<1>(2l) * GroupT::one();
+        P.to_affine_coordinates();
+        using Fq = typename std::decay<decltype(P.X)>::type;
+        Fq LHS = (P.Y * P.Y);
+        Fq RHS =
+            ((P.X * P.X * P.X) + (GroupT::coeff_a * P.X) + GroupT::coeff_b);
+        ASSERT_EQ(LHS, RHS);
+    }
+    // P = random
+    {
+        GroupT P = GroupT::random_element();
+        P.to_affine_coordinates();
+        using Fq = typename std::decay<decltype(P.X)>::type;
+        Fq LHS = (P.Y * P.Y);
+        Fq RHS =
```

Comment:
Few things here:

1) could you make this function generic over the group type `GroupT`.  It seems to only test G1.

2) We can factor out the check either into a separate function or a local function, reducing all this to something like:
```
check_curve_equation(GroupT::one());
check_curve_equation(bigint<1>(-1l) * GroupT::one());
check_curve_equation(bigint<1>(2l) * GroupT::one());
check_curve_equation(GroupT::random_element());
```

3) Note, `bigint` here can be replaced by `Fr` (where we may need something like `using Fr = GroupT::scalar_field`.

Reply:
Good suggestions -- thanks! All done.